Step3

oyj7677 opened the pull request:
테스트 코드 함수의 네이밍이 어려워 주석으로 대체하였습니다.
테스트 코드 주석 작성을 given, when, then을 사용해봤습니다.
 - given의 경우 변수 저장을 따로 하지 않아 사용하지 않았습니다.
 - given, when, then 을 적절하게 사용했는지 알고싶습니다.
 - 에러검증의 경우 assertEquals를 사용하지 않고  @Test(expected = IllegalArgumentException::class)와 같은 식으로 사용하는 것이 맞나요?

이상입니다. 감사합니다.

Commit:
feat: 사용자가 피연산자 0 ~ 9 버튼을 누르면 화면에 해당 숫자를 보여준다

Commit:
Merge branch 'main' of https://github.com/oyj7677/android-calculator-mvp

Commit:
step3: 계산기 도메인과 관련된 로직은 domain 모듈에 구현한다.

oyj7677 commented:
step3 과제 제출입니다. 잘부탁드립니다.

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,59 @@
+        calculator = Calculator()
+    }
+
+    @Test
+    fun operatorTest1() {
+        // when : calculator를 통해 수식을 계산한다.
+        val actual = calculator.evaluate("1 + 2 + 3")
+
+        // then : 사칙 연산의 우선 순위가 아닌 입력 우선 순위로 계산을 진행한다.
+        assertEquals(6, actual)
```
테스트 메서드 명에 테스트하려는 내용을 한글로 적어보는 건 어떨까요?
주석에 달아둘 내용을 메서드명으로 충분히 나타낼 수 있어 보이네요 :)

아래는 예시이니 참고로 활용해주세요!
`문자열 사칙 연산을 수행한다`

malibinYun replied:
또, assertEquals를 활용하기보다, Truth를 활용해보는 건 어떨까요?

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,59 @@
+    @Test
+    fun operatorTest2() {
+        // when : calculator를 통해 수식을 계산한다.
+        val actual = calculator.evaluate("3 * 2 + 8 / 3")
+
```
테스트하고 싶은 수식이 여러 개라면, Parameterized Test를 활용해보는 건 어떨까요?

malibinYun replied:
> given, when, then 을 적절하게 사용했는지 알고싶습니다.

극단적으로 분리해본다면, 주어진 수식을 지역 변수로 분리해 given으로도 둘 수 있습니다 :)
지금은 적절히 잘 나누어주셨네요! 👍

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,59 @@
+        assertThat(actual).isEqualTo(4)
+    }
+
+    @Test(expected = IllegalArgumentException::class)
+    fun operatorErrorTest1() {
+        // when : 수식을 공백 문자열로 채운다.
+        calculator.evaluate("     ")
+
+        // then : IllegalArgumentException이 발생한다.
+    }
```
>에러검증의 경우 assertEquals를 사용하지 않고 `@test(expected = IllegalArgumentException::class)`와 같은 식으로 사용하는 것이 맞나요?

이렇게도 ExceptionTest를 작성해볼 수 있겠네요!

exception이 발생할 메서드를 runCatching으로 감싸고, 어떤 Exception 인스턴스가 만들어졌는지, 어떤 에러메시지가 나왔는지 검증해보는 건 어떨까요?
아래는 참고용으로 활용해주세요!

```kotlin
val exception = runCatching { calculator.evalate(" ") }.exceptionOrNull()

assertThat(ex).isInstanceOf(~)
assertThat(ex).hasMessageThat().isEqualTo("~")
```

malibinYun replied:
또, whiteSpace에 해당하는 공백 문자들을 ParameterizedTest로 테스트해보는 건 어떨까요?
"\t", "\n" 등 여러 공백 문자가 있을 수 있겠네요 :)

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+
+class Calculator {
+    fun evaluate(str: String?): Int {
+        val numAndOperator = str?.split(" ") ?: throw IllegalArgumentException()
+        if(str.trim() == "")  throw IllegalArgumentException()
+
```
어떤 이유에 Exception이 발생했는 지 메시지를 담아보는 건 어떨까요?
디버깅할 때 유용해집니다 :)

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+
+            if(!operatorVerification(operator)) throw IllegalArgumentException()
+
+            when(numAndOperator[i]) {
+                "+" -> result = addition(result, num.toInt())
+                "-" -> result = subtraction(result, num.toInt())
+                "*" -> result = multiplication(result, num.toInt())
+                "/" -> result = division(result, num.toInt())
+            }
```
사칙 연산 기호를 enum class로 분리해보는 건 어떨까요?

malibinYun reviewed (changes requested):
3단계 미션 고생 많으셨어요!
생각해볼 만한 코멘트 남겨두었습니다.
피드백 반영 후 다시 리뷰 요청 부탁드려요!

Commit:
1차 피드백에 따른 변경사항입니다.
1. 테스트 메서드 명 한글로 수정
2. assertEquals ->  assertThat로 수정
3. 에러 검증 시 runCatching를 통해 에러 검증
4. Exception의 발생원인을 주석으로 추가하였습니다.
5. 사칙연산 기호를 enum class로 분리하였습니다.

Commit:
마지막 에러 테스트에서 runCatching 후 .exceptionOrNull()포함을 깜박하고 푸시하였습니다.

Commit:
정상케이스에 있어 Parameterized Test를 추가하였습니다.

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,94 @@
+import org.junit.runners.Parameterized
+import org.junit.runners.Parameterized.Parameters
+
+@RunWith(value = Parameterized::class)
+class CalculatorTest(val input: String, val expected: Int) {
```
Parameterized Test를 작성해주셨네요! 👍
Junit4의 ParameterizedTest는 클래스 단위로 해야하는 단점이 있지요.
이 멤버 변수들이 아래 테스트 목록에 모두 영향을 받는 것이 아니기에, 테스트 가독성을 떨어뜨릴 수 있어요.

kotlin에는 한 파일에 여러 class를 작성할 수 있으니, ParameterizedTest해야하는 각 항목 마다 클래스로 분리해보는 건 어떨까요?
동일한 멤버를 사용하는 경우에는 한 클래스로 묶어볼 수도 있겠네요 :)

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,94 @@
+
+    @Test
+    fun 문자열_수식이_공백일_경우_에러를_발생시킨다() {
+        // when : 수식을 공백 문자열로 채운다.
+        val exception = kotlin.runCatching { calculator.evaluate("        ") }.exceptionOrNull()
```
`kotlin.`은 제거하셔도 괜찮습니다 :)
```suggestion
        // when : 수식을 공백 문자열로 채운다.
        val exception = runCatching { calculator.evaluate("        ") }.exceptionOrNull()
```

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Calculator.kt:
```diff
@@ -0,0 +1,56 @@
+
+class Calculator {
+    fun evaluate(str: String?): Int {
+        // 문자열 수식(str)이 null일 경우 IllegalArgumentException 발행
+        val numAndOperator = str?.split(" ") ?: throw IllegalArgumentException()
```
https://github.com/next-step/android-calculator-mvp/pull/196#discussion_r1278302646

Exception 객체를 생성할 때 에러 메시지를 담아보라는 의미었습니다 :)
아래는 참고로 활용해주세요!
```suggestion
        // 문자열 수식(str)이 null일 경우 IllegalArgumentException 발행
        val numAndOperator = str?.split(" ") ?: throw IllegalArgumentException("입력 수식이 null이거나 빈 문자열이어서는 안됨. 입력 값 : $str")
```

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Calculator.kt:
```diff
@@ -0,0 +1,56 @@
+        // 문자열 수식(str)이 null일 경우 IllegalArgumentException 발행
+        val numAndOperator = str?.split(" ") ?: throw IllegalArgumentException()
+
+        // 문자열 수식(str)이 공백일 경우 IllegalArgumentException 발행
+        if(str.trim() == "")  throw IllegalArgumentException()
```
require() 를 활용해보는 건 어떨까요?
또, String.isEmpty() 확장함수를 활용해볼 수도 있겠네요 :)

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Calculator.kt:
```diff
@@ -0,0 +1,56 @@
+
+            if(!operatorVerification(operator)) throw IllegalArgumentException()
+
+            when(operator) {
+                Operator.PLUS.operator -> result = addition(result, num.toInt())
+                Operator.MINUS.operator -> result = subtraction(result, num.toInt())
+                Operator.MULTIPLIED.operator -> result = multiplication(result, num.toInt())
+                Operator.DIVIDED.operator -> result = division(result, num.toInt())
+            }
+        }
```
enum class 에서 문자열을 받아서 해당 Operator 객체를 반환하는 메서드를 만들어보는 건 어떨까요?
아래 코드의 형태 처럼 작성할 수 있게 됩니다.
```kotlin
val operator = Operator.find(numAndOperator[i])

when(operator) {
    Operator.PLUS -> 
    Operator.MINUS -> ...
...
}
```

그 다음으로는 해당 Operator에 대한 계산 전략을 enum이 직접 갖게 해보는 것도 시도해보시면 좋겠어요 :)
그럼 아래 형태 처럼 when절을 없애는 방향으로 작성해볼 수 있습니다.
```kotlin
val operator = Operator.find(numAndOperator[i])
val result = operator.calculate(leftNumber, rightNumber)
```

malibinYun reviewed (approved):
피드백 반영하시느라 고생 많으셨어요 :)
고민해보면 좋을 점들에 코멘트 남겨두었습니다.
피드백은 다음 미션에 반영해주세요!
